crypto/context.c

```diff
@@ -117,3 +135,56 @@ void *openssl_ctx_get_data(OPENSSL_CTX *ctx, int index)
     return data;
 }
 
+int openssl_ctx_run_once(OPENSSL_CTX *ctx, unsigned int idx,
+                         openssl_ctx_run_once_fn run_once_fn)
+{
+    int done = 0, ret = 0;
+
+#ifndef FIPS_MODE
+    if (ctx == NULL) {
+        if (!RUN_ONCE(&default_context_init, do_default_context_init))
+            return 0;
+        ctx = &default_context;
+    }
+#endif
+    if (ctx == NULL)
+        return 0;
+
+    CRYPTO_THREAD_read_lock(ctx->lock);
+    done = ctx->run_once_done[idx];
+    if (done)
+        ret = ctx->run_once_ret[idx];
```

Comment:
You don't really need to get the stored return value within guard of the lock...  If `done` is true, the stored return value is constant, period.  If not, this implementation can't be right.